README
NAME
    MooX::ClassAttribute - declare class attributes Moose-style... but without
    Moose

SYNOPSIS
       {
          package Foo;
          use Moo;
          use MooX::ClassAttribute;
          class_has ua => (
             is      => 'rw',
             default => sub { LWP::UserAgent->new },
          );
       }
   
       my $r = Foo->ua->get("http://www.example.com/");

DESCRIPTION
    This module adds support for class attributes to Moo. Class attributes are
    attributes whose values are not associated with any particular instance of
    the class.

    For example, the `Person` class might have a class attribute
    "binomial_name"; its value "Homo sapiens" is not associated with any
    particular individual, but the class as a whole.

       say Person->binomial_name;   # "Homo sapiens"
       my $bob = Person->new;
       say $bob->binomial_name;     # "Homo sapiens"
   
       my $alice = Person->new;
       $alice->binomial_name("H. sapiens");
       say $bob->binomial_name;     # "H. sapiens"

    Class attributes may be defined in roles, however they cannot be called as
    methods using the role package name. Instead the role must be composed
    with a class; the class attributes will be installed into that class.

    This module mostly tries to behave like MooseX::ClassAttribute.

CAVEATS
    *   Overriding class attributes and their accessors in subclasses is not
        yet supported. The implementation, and expected behaviour hasn't been
        figured out yet.

    *   When Moo classes are inflated to Moose classes, this module will
        *attempt* to load MooseX::ClassAttribute, and use that to provide
        class attribute meta objects.

        If MooseX::ClassAttribute cannot be loaded, a loud warning will be
        printed, and the inflation will fall back to representing class
        attribute accessors as plain old class methods.

    *   This module uses some pretty experimental techniques, especially to
        handle inflation. There are probably all sorts of bugs lurking. Don't
        let that scare you though; I'm usually pretty quick to fix bugs once
        they're reported. ;-)

BUGS
    Please report any bugs to
    <http://rt.cpan.org/Dist/Display.html?Queue=MooX-ClassAttribute>.

    See also: "CAVEATS" in Method::Generate::ClassAttribute.

SEE ALSO
    Moo, MooseX::ClassAttribute.

AUTHOR
    Toby Inkster <tobyink@cpan.org>.

COPYRIGHT AND LICENCE
    This software is copyright (c) 2013 by Toby Inkster.

    This is free software; you can redistribute it and/or modify it under the
    same terms as the Perl 5 programming language system itself.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
MooX-ClassAttribute
===================

Created:      2012-12-27
Home page:    <https://metacpan.org/release/MooX-ClassAttribute>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=MooX-ClassAttribute>
Maintainer:   Toby Inkster (TOBYINK) <tobyink@cpan.org>

0.010	2013-09-06

 [ Bug Fixes ]
 - Skip test case that requires Moo 1.002000 on older versions of Moo.
   Fixes RT#88459.
   Andreas J. König++
   <https://rt.cpan.org/Ticket/Display.html?id=88459>

 [ Documentation ]
 - Correct link to Method::Generate::ClassAccessor in documentation.

 [ Packaging ]
 - Conflict detection for Moo == 1.001000 in Makefile.PL.

0.009	2013-08-27

 [ Bug Fixes ]
 - Fix MooX::CaptainHook's `on_application` hook to work with recent
   versions of Moo::Role/Role::Tiny which do not always call
   `apply_single_role_to_package`.
   Graham Knop++

0.008	2013-07-10

 [ Bug Fixes ]
 - Support non-coderef defaults.
   Fixes RT#87638.
   Rob Bloodgood++
   <https://rt.cpan.org/Ticket/Display.html?id=87638>

 [ Packaging ]
 - Switch to Dist::Inkt.

0.007	2013-07-10

 [ Bug Fixes ]
 - Fixed error: Can't call method "isa" on an undefined value at
   MooX/CaptainHook.pm line 27.
   Fixes RT#86828.
   Dinis Rebolo++
   <https://rt.cpan.org/Ticket/Display.html?id=86828>

 [ Documentation ]
 - Note incompatibility with Moo 1.001000.

0.006	2013-01-11

 [ Bug Fixes ]
 - Avoid triggering an 'in cleanup' error on some older versions of Perl.

0.005	2013-01-05

 [ Bug Fixes ]
 - Avoid triggering Sub::Exporter::Progressive's dependency on
   Sub::Exporter.

0.004	2013-01-03

 [ Bug Fixes ]
 - Fix MooX::CaptainHook on_inflation fragility when Moose is loaded early.

0.003	2013-01-03

 [ Bug Fixes ]
 - Prevent MooX::CaptainHook from inadvertantly loading Moose.

0.002	2013-01-01

 [ Packaging ]
 - List dependencies.

0.001	2013-01-01	Initial release


Makefile.PL
use strict;
use ExtUtils::MakeMaker 6.17;

my $EUMM = eval( $ExtUtils::MakeMaker::VERSION );

my $meta = {
  "abstract"       => "declare class attributes Moose-style... but without Moose",
  "author"         => ["Toby Inkster (TOBYINK) <tobyink\@cpan.org>"],
  "dynamic_config" => 0,
  "generated_by"   => "Dist::Inkt::Profile::TOBYINK version 0.009, CPAN::Meta::Converter version 2.120921",
  "keywords"       => [],
  "license"        => ["perl_5"],
  "meta-spec"      => {
                        url => "http://search.cpan.org/perldoc?CPAN::Meta::Spec",
                        version => 2,
                      },
  "name"           => "MooX-ClassAttribute",
  "no_index"       => { directory => ["eg", "examples", "inc", "t", "xt"] },
  "prereqs"        => {
                        configure => {
                                       requires => {
                                         "CPAN::Meta::Requirements" => "2.000",
                                         "ExtUtils::MakeMaker"      => 6.17,
                                       },
                                     },
                        develop   => {
                                       recommends => { "Dist::Inkt" => 0 },
                                       requires   => { "Moose" => 0, "MooseX::ClassAttribute" => 0, "Sub::Exporter" => 0 },
                                     },
                        runtime   => {
                                       conflicts  => { Moo => "== 1.001000" },
                                       recommends => { "MooseX::ClassAttribute" => 0 },
                                       requires   => {
                                                       "Moo" => "1.000000",
                                                       "Role::Tiny" => "1.000000",
                                                       "Sub::Exporter::Progressive" => 0,
                                                     },
                                     },
                      },
  "provides"       => {
                        "Method::Generate::ClassAccessor"         => { file => "lib/Method/Generate/ClassAccessor.pm", version => "0.010" },
                        "MooX::CaptainHook"                       => { file => "lib/MooX/CaptainHook.pm", version => "0.010" },
                        "MooX::CaptainHook::HandleMoose::Hack"    => { file => "lib/MooX/CaptainHook.pm", version => "0.010" },
                        "MooX::CaptainHook::OnApplication"        => { file => "lib/MooX/CaptainHook.pm", version => "0.010" },
                        "MooX::CaptainHook::OnApplication::Moose" => { file => "lib/MooX/CaptainHook.pm", version => "0.010" },
                        "MooX::CaptainHook::OnInflation"          => { file => "lib/MooX/CaptainHook.pm", version => "0.010" },
                        "MooX::ClassAttribute"                    => { file => "lib/MooX/ClassAttribute.pm", version => "0.010" },
                        "MooX::ClassAttribute::HandleMoose"       => {
                                                                       file => "lib/MooX/ClassAttribute/HandleMoose.pm",
                                                                       version => "0.010",
                                                                     },
                      },
  "release_status" => "stable",
  "resources"      => {
                        bugtracker   => {
                                          web => "http://rt.cpan.org/Dist/Display.html?Queue=MooX-ClassAttribute",
                                        },
                        homepage     => "https://metacpan.org/release/MooX-ClassAttribute",
                        license      => ["http://dev.perl.org/licenses/"],
                        repository   => {
                                          type => "git",
                                          web  => "https://github.com/tobyink/p5-moox-classattribute",
                                        },
                        X_identifier => "http://purl.org/NET/cpan-uri/dist/MooX-ClassAttribute/project",
                      },
  "version"        => "0.010",
  "x_contributors" => ["Dinis Rebolo (DREBOLO) <drebolo\@cpan.org>"],
};

my %dynamic_config;
for my $stage (keys %{$meta->{prereqs}})
{
	my $conflicts = $meta->{prereqs}{$stage}{conflicts} or next;
	require CPAN::Meta::Requirements;
	$conflicts = 'CPAN::Meta::Requirements'->from_string_hash($conflicts);
	
	for my $module ($conflicts->required_modules)
	{
		eval "require $module" or next;
		my $installed = eval(sprintf('$%s::VERSION', $module));
		$conflicts->accepts_module($module, $installed) or next;
		
		my $message = "\n".
			"** This version of $meta->{name} conflicts with the version of\n".
			"** module $module ($installed) you have installed.\n";
		die($message . "\n" . "Bailing out")
			if $stage eq 'build' || $stage eq 'configure';
		
		$message .= "**\n".
			"** It's strongly recommended that you update it after\n".
			"** installing this version of $meta->{name}.\n";
		warn("$message\n");
	}
}

my %WriteMakefileArgs = (
	ABSTRACT   => $meta->{abstract},
	AUTHOR     => ($EUMM >= 6.5702 ? $meta->{author} : $meta->{author}[0]),
	DISTNAME   => $meta->{name},
	VERSION    => $meta->{version},
	EXE_FILES  => [ map $_->{file}, values %{ $meta->{x_provides_scripts} || {} } ],
	NAME       => do { my $n = $meta->{name}; $n =~ s/-/::/g; $n },
	%dynamic_config,
);

$WriteMakefileArgs{LICENSE} = $meta->{license}[0] if $EUMM >= 6.3001;

sub deps
{
	my %r;
	for my $stage (@_)
	{
		for my $dep (keys %{$meta->{prereqs}{$stage}{requires}})
		{
			my $ver = $meta->{prereqs}{$stage}{requires}{$dep};
			$r{$dep} = $ver if !exists($r{$dep}) || $ver >= $r{$dep};
		}
	}
	\%r;
}

my ($build_requires, $configure_requires, $runtime_requires, $test_requires);
if ($EUMM >= 6.6303)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{TEST_REQUIRES}      ||= deps('test');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');
}
elsif ($EUMM >= 6.5503)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build', 'test');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');	
}
elsif ($EUMM >= 6.52)
{
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime', 'build', 'test');	
}
else
{
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('configure', 'build', 'test', 'runtime');	
}

{
	my $minperl = delete $WriteMakefileArgs{PREREQ_PM}{perl};
	exists($WriteMakefileArgs{$_}) && delete($WriteMakefileArgs{$_}{perl})
		for qw(BUILD_REQUIRES TEST_REQUIRES CONFIGURE_REQUIRES);
	if ($minperl and $EUMM >= 6.48)
	{
		$WriteMakefileArgs{MIN_PERL_VERSION} ||= $minperl;
	}
	elsif ($minperl)
	{
		die "Need Perl >= $minperl" unless $] >= $minperl;
	}
}

sub FixMakefile
{
	return unless -d 'inc';
	my $file = shift;
	
	local *MAKEFILE;
	open MAKEFILE, "< $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;
	
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1'inc', /;
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ -Iinc$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ "-Iinc"$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 "-Iinc"/m;
	$makefile =~ s/^(PERL = .*)/$1 "-Iinc"/m;
	
	open  MAKEFILE, "> $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	print MAKEFILE $makefile or die $!;
	close MAKEFILE or die $!;
}

my $mm = WriteMakefile(%WriteMakefileArgs);
FixMakefile($mm->{FIRST_MAKEFILE} || 'Makefile');
exit(0);

